src/client.rs

```diff
@@ -18,20 +19,22 @@ impl Service for Client {
     type Resp = String;
     type Error = io::Error;
     // Again for simplicity, we are just going to box a future
-    type Fut = Box<Future<Item = Self::Resp, Error = io::Error>>;
+    type Fut = Box<Future<Item = Self::Resp, Error = io::Error> + Send>;
 
     fn call(&self, req: String) -> Self::Fut {
         self.inner.call(pipeline::Message::WithoutBody(req))
             .boxed()
     }
 }
 
-pub fn connect(reactor: ReactorHandle, addr: &SocketAddr) -> io::Result<Client> {
-    let addr = addr.clone();
-    let client = pipeline::connect(&reactor, move || {
-        let stream = try!(TcpStream::connect(&addr));
-        Ok(new_line_transport(stream))
-    });
+pub fn connect(loop_handle: LoopHandle, addr: &SocketAddr) -> IoFuture<Client> {
+    let h2 = loop_handle.clone();
+    loop_handle.tcp_connect(addr)
+        .and_then(move |stream| {
+            let c = pipeline::connect(h2, Take::new(move || {
```

Comment:
this got way more complicated too - why is Take needed? Avoiding it would make this easier to understand.